[D2] NavigationBar 구현

kimkyunghun3 opened the pull request:
# 개요
- 🔗  이슈링크 : https://github.com/mash-up-kr/GGiriGGiri_iOS/issues/5


# 변경사항
## 작업 내용
UINavigationBar 구현
ButtonTapEvent 발생 시 어떤 버튼을 누른 것인지 알고 그에 대한 이벤트가 발생되었다고 전송되도록 구현

색깔, 이미지, 폰트는 모두 임의로 넣은것입니다!

구현하면서 rightBarButtonItem이 2개이거나 없는 상황을 유연하게 대처하기위해 배열을 만들고 있는지 여부에 따라 추가하거나 추가하지 않는 로직이 생기는 문제가 발생했습니다..ㅜㅜ
NavigationBar이지만 UIView로 구현해서 이곳에 로직이 있는게 조금 애매한거같은데 어떻게 해야할까요?

### 스크린샷
<img src = "https://user-images.githubusercontent.com/52434820/177139097-1b59371c-4d4a-43fc-8de3-28dd7bb14000.png" width = "300">

<img src = "https://user-images.githubusercontent.com/52434820/177139106-58988c95-a472-4e82-846a-07fdabbc718f.png" width ="300">

<img src = "https://user-images.githubusercontent.com/52434820/177139121-d6605bbb-176c-4d8d-830d-e8e208f97264.png" width ="300">

kimkyunghun3 committed:
:bulb: NavigationBar 구현

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

szzang-kr commented on Projects/DesignSystem/Sources/Component/NavigationBar/DDIPNavigationBar.swift:
```diff
@@ -0,0 +1,142 @@
+        let navigationBar = UINavigationBar(frame: CGRect(x: 0, y: 56, width: UIScreen.main.bounds.size.width, height: 38))
+        let navigationItem = UINavigationItem(title: self.title ?? "")
+        
+        if let alarm = alarm, let mybox = mybox {
+            rigtBarButtons.append(alarm)
+            rigtBarButtons.append(mybox)
+        }
```
여기가 문제인것 같은데, 컴포넌트에서 버튼을 직접 추가하지 않고.

아래 예시를 참고해서 
사용처에서 버튼을 추가할 수 있도록 로직을 변경하면 위에서 작성한 로직이 필요 없을 것 같아요.


```swift
// DDIPNavigationBar.swift
private var buttons: [UIBarButtonItem] = []

init(..., ..., buttons: [UIButtonItems]) {
  self.buttons = buttons
}

private func setNavigationBar() {
    ....
    ....
    navigationBar.setItems(buttons, animated: false)
}

// Anywhere
let alarmButton = UIButtonItems(...)
let couponButton = UIButtonItems(...)

// make NavigationBar with ButtonItems
DDIPNavigationBar(..., ..., [
    alarmButton,
    couponButton
])

// make tapAction with closure or Rx
alarmButton.tapAction {
    print("tapped couponButton")
}

couponButton.rx.tap
  .subsribeOn(onNext: { 
      print("tapped couponButton")
  })
  .disposed(by: disposeBag)

kimkyunghun3 replied:
말씀해주신 방향으로 수정해보았습니다. 확실히 `Component`로 사용되는 `Button`만 설정해주고 그 외에 보편적으로 사용하지 않는 것은 사용처에서 그 상황에서 만드는 것이 더 좋은 방향인거 같습니다!  좋은 피드백 감사합니다 👍 
`rightButtonsItems`는 사용처에서 만들어서 넣으면 `NavigationBar` 오른쪽에 들어가도록만 설정해두었습니다!
```swift
private var rightButtonsItem: [UIBarButtonItem]?
.... 
navigationItem.rightBarButtonItems = rightButtonsItem

// 사용처
let ddipnavibar = DDIPNavigationBar(leftBarItem: .cancel, title: "hello", rightButtonsItem: [
    alarmButton, couponButton
])
```

szzang-kr commented on Projects/DesignSystem/Sources/Component/NavigationBar/DDIPNavigationBar.swift:
```diff
@@ -0,0 +1,142 @@
+    }
+    
+    private func bind() {
+        back?.rx.tap.bind(to: self.backButtonTapEvent)
+            .disposed(by: disposeBag)
+        mybox?.rx.tap
+            .bind(to: self.myboxButtonTapEvent)
+            .disposed(by: disposeBag)
+        alarm?.rx.tap.bind(to: self.alarmButtonTapEvent)
+            .disposed(by: disposeBag)
+    }
```
하지만 back button, close button은 모든 화면에서 공통으로 쓰일 수 있으니까 component에서 지정하는 편이 좋을 것 같네요!

kimkyunghun3 replied:
위에 댓글에 설명했습니다. 이부분만 남기고 나머지는 제거하도록 했습니다 :)

szzang-kr commented on Projects/DesignSystem/Sources/Component/NavigationBar/DDIPNavigationBar.swift:
```diff
@@ -0,0 +1,142 @@
+import RxCocoa
+
+public class DDIPNavigationBar: UIView {
+    public enum BarItem {
+        case mybox
+        case noAlarm
+        case alarm
+        case back
+        case cancel
+        
+        var icon: UIImage? {
+            switch self {
+            case .mybox:
+                return UIImage(systemName: "swift")
+            case .noAlarm:
+                return UIImage(systemName: "bell")
+            case .alarm:
+                return UIImage(systemName: "bell.badge")
+            case .back:
+                return UIImage(systemName: "chevron.backward")
+            case .cancel:
+                return UIImage(systemName: "xmark")
+            }
+        }
+    }
```
컴포넌트에서 사용될 case를 지정해두는것은 좋은 생각이나, 컴포넌트에서 사용될 case가 많아지면 `enum`도 겉잡을 수 없이 커지게됩니다.

따라서 특정화면에서만 사용되는 (mybox, noAlarm, alarm) 같은 경우는 사용하는 화면에서 삽입 할 수 있도록 하는 편이 좋습니다.

component는 모든 화면에서 쓰이기 때문에 가장 보편적이여야 합니다.

kimkyunghun3 replied:
넵 사용해보니 점점 늘어나는 `enum`에서 관리가 어려울 뿐만 아니라 재사용성이 오히려 더 떨어질 것같다는 판단을 하게 되었습니다. 좋은 제안 감사합니다! 사용처를 더 고려해서 구현해보겠습니다!

szzang-kr reviewed (changes requested):
컴포넌트는 만드는것보다 사용하는 사람을 더욱 생각해야합니다.

직접 사용해 보고 문제점을 파악하는것도 좋은 도움이 될 것 이라고 생각합니다.
수고하셨습니다.  👍

AhnSangHee commented on Projects/DesignSystem/Sources/Component/NavigationBar/DDIPNavigationBar.swift:
```diff
@@ -0,0 +1,142 @@
+    }
+    
+    private func setNavigationBar() {
+        var rigtBarButtons: [UIBarButtonItem] = []
```
```suggestion
        var rightBarButtons: [UIBarButtonItem] = []
```
🔴 오타가 있네요!

kimkyunghun3 replied:
오타 지적 감사합니다! 네이밍 수정 했습니다!

AhnSangHee reviewed (changes requested):
수고하셨습니다👍

kimkyunghun3 committed:
:recycle: rightButtonItem 네이밍 오타 수정

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

kimkyunghun3 committed:
:recycle: component NavigationBar에 back, cancel만 넣도록 변경 #5

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################

szzang-kr commented on Projects/DesignSystem/Sources/Component/NavigationBar/DDIPNavigationBar.swift:
```diff
@@ -35,28 +26,23 @@ public class DDIPNavigationBar: UIView {
-    private var back: UIBarButtonItem?
-    private var mybox: UIBarButtonItem?
-    private var alarm: UIBarButtonItem?
+    private var rightButtonsItem: [UIBarButtonItem]?
```
`rightButtonsItems` 네이밍이 더 좋을것 같습니다. ~Item이라서 `UIBarButtonItem` 단일 버튼인 줄 알았습니다. 😓

kimkyunghun3 replied:
말씀해주신 방향으로 네이밍 변경했습니다!

szzang-kr commented on Projects/DesignSystem/Sources/Component/NavigationBar/DDIPNavigationBar.swift:
```diff
@@ -35,28 +26,23 @@ public class DDIPNavigationBar: UIView {
-    public let backButtonTapEvent = PublishRelay<Void>()
-    public let myboxButtonTapEvent = PublishRelay<Void>()
-    public let alarmButtonTapEvent = PublishRelay<Void>()
+    public let leftButtonTapEvent = PublishRelay<Void>()
```
이 relay의 역할은 무엇일까요..?

kimkyunghun3 replied:
설계 목적은 이벤트가 발생했을 떄 사용처에 이벤트가 발생했다고 전달해주는 역할로 생각했습니다. UI event에서 이벤트가 발생했다 정도만 알려주면 된다고 생각해서 이렇게 했습니다!

szzang-kr replied:
발생한 이벤트를 어디에 전달해주는것이 목적일까요?

kimkyunghun3 replied:
제 생각에는 ViewModel에 전달해줘야 된다고 생각했습니다! 그래서 그곳에서 만약 화면전환이나 다른 action를 해야한다면 그곳에서 처리하면된다고 생각했습니다.

szzang-kr replied:
아 헉 public이군요 죄송합니다. private으로 착각했습니다 😭

szzang-kr commented on Projects/DesignSystem/Sources/Component/NavigationBar/DDIPNavigationBar.swift:
```diff
@@ -65,78 +51,44 @@ public class DDIPNavigationBar: UIView {
-        get { self.back?.image }
-        set { self.back?.setBackgroundImage(newValue, for: .normal, barMetrics: .default) }
+        get { self.leftButtonItem?.image }
+        set { self.leftButtonItem?.setBackgroundImage(newValue, for: .normal, barMetrics: .default) }
```
left 버튼은 close, back 두가지 상황밖에 없어서 component에서 지정하면 좋을 것 같은데..

이것도 사용처에서 지정하도록 해두신걸까요?

참조 : https://github.com/mash-up-kr/GGiriGGiri_iOS/pull/22#discussion_r914356173

kimkyunghun3 replied:
제 생각에서는 이를 지정해줄수 있긴 한데, 그러면 Component로 두가지 케이스를 대응해주지 못한다고 생각했습니다. 그래서 이렇게 사용처에서 지정해주도록 생각했는데 혹시 다른 방향을 이야기하시는 것일까요?

szzang-kr replied:
right 버튼의 경우 특정 상황에서만 사용되는 반면에, left 버튼의 경우는 많은곳에서 동일하게 사용됩니다.

사용처에서 해당 버튼을 추가하게되면 동일한 코드가 많은곳에서 작성될 것 같습니다.

kimkyunghun3 replied:
아하 그럼 제가 이해한바로는 하나의 네비게이션바 파일에서 두개의 init를 만들어서 각각 다른 버튼을 초기화된 상태로 만드는 것을 만들면 되겠다 라고 생각했는데 이 방식으로 하면 괜찮지 않을까요?

szzang-kr replied:
제가 잘못 보고 있었습니다!! 이대로 진행해도 괜찮습니다!

szzang-kr reviewed (changes requested):
리뷰가 늦어서 죄송합니다. 😭

kimkyunghun3 committed:
:recycle: rightButtonsItems로 네이밍 변경

# 제목을 아랫줄에 작성, 제목 끝에 마침표 금지, 무엇을 했는지 명확하게 작성

################
# 세부 내용을 아랫줄에 작성

################
# 추가적인 세부 내용을 아랫줄에 작성

################
# :tada: = 프로젝트 시작
# :sparkles: = create 파일 생성 시
# :bulb: = update 코드 및 UI 업데이트, 기능 구현 시
# :rocket: = release 배포 시
# :recycle: = refactor 리팩토링 시 (기능은 같은데 코드가 달라지는경우)
# :memo: = 문서 작성 시
# :art: = 디자인 시스템 관련 내용 (Color / Asset)
# :hammer: = 테스트 코드 작성
################
